File: Year-2023/LeetCode-March-Challenge/NumberofOperationstoMakeNetworkConnected.java
import java.util.*;

class Solution {
    private void DFS(ArrayList<Integer>[] adj, boolean[] visited, int node){
        visited[node] = true;
        for(int x: adj[node]){
            if(!visited[x]){
                DFS(adj, visited, x);
            }
        }
        return;
    }
    public int makeConnected(int n, int[][] connections) {
        if (connections.length < n - 1)
            return -1;
        ArrayList<Integer>[] adj = new ArrayList[n];
        for (int i = 0; i < n; ++i) {
            adj[i] = new ArrayList<>();
        }

        for (int[] x : connections) {
            adj[x[0]].add(x[1]);
            adj[x[1]].add(x[0]);
        }
        boolean[] visited = new boolean[n];
        Arrays.fill(visited, false);
        int count = 0;
        for(int i = 0; i < n; ++i){
            if(!visited[i]){
                DFS(adj, visited, i);
                count++;
            }
        }
        return count - 1;
    }
}

public class NumberofOperationstoMakeNetworkConnected {
    public static void main(String[] args) {
        int n = 4; int[][]connections = {{0,1},{0,2},{1,2}};
        Solution s = new Solution();
        System.out.println(s.makeConnected(n, connections));
    }
}

File: Year-2023/LeetCode-March-Challenge/NumberofOperationstoMakeNetworkConnected.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
private:
    void DFS(vector<vector<int>> &adj, vector<bool> &visited, int node){
        visited[node] = true;
        for(auto x: adj[node]){
            if(!visited[x]){
                DFS(adj, visited, x);
            }
        }
        return;
    }
public:
    int makeConnected(int n, vector<vector<int>>& connections) {
        if(connections.size() < n - 1) return -1;
        vector<vector<int>> adj(n);
        for(auto x: connections){
            adj[x[0]].push_back(x[1]);
            adj[x[1]].push_back(x[0]);
        }
        vector<bool> visited(n, false);
        int count = 0;
        for(int i = 0; i < n; ++i){
            if(!visited[i]){
                DFS(adj, visited, i);
                count++;
            }
        }
        return count - 1;
    }
};

int main(){
    int n = 4; vector<vector<int>> connections = {{0,1},{0,2},{1,2}};
    Solution *s = new Solution();
    cout<<s->makeConnected(n, connections)<<endl;
    return 0;
}